bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs

```diff
@@ -0,0 +1,270 @@
+#[derive(PartialEq, Debug)]
+pub enum XcmConverterError {
+	UnexpectedEndOfXcm,
+	EndOfXcmMessageExpected,
+	WithdrawAssetExpected,
+	DepositAssetExpected,
+	NoReserveAssets,
+	FilterDoesNotConsumeAllAssets,
+	TooManyAssets,
+	ZeroAssetTransfer,
+	BeneficiaryResolutionFailed,
+	AssetResolutionFailed,
+	InvalidFeeAsset,
+	SetTopicExpected,
+	ReserveAssetDepositedExpected,
+	InvalidAsset,
+	UnexpectedInstruction,
+	TooManyCommands,
+	AliasOriginExpected,
+	InvalidOrigin,
+	TransactDecodeFailed,
+	TransactParamsDecodeFailed,
+	FeeAssetResolutionFailed,
+	CallContractValueInsufficient,
+}
+
+macro_rules! match_expression {
+	($expression:expr, $(|)? $( $pattern:pat_param )|+ $( if $guard: expr )?, $value:expr $(,)?) => {
+		match $expression {
+			$( $pattern )|+ $( if $guard )? => Some($value),
+			_ => None,
+		}
+	};
+}
+
+pub struct XcmConverter<'a, ConvertAssetId, Call> {
+	iter: Peekable<Iter<'a, Instruction<Call>>>,
+	ethereum_network: NetworkId,
+	_marker: PhantomData<ConvertAssetId>,
+}
+impl<'a, ConvertAssetId, Call> XcmConverter<'a, ConvertAssetId, Call>
+where
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+{
+	pub fn new(message: &'a Xcm<Call>, ethereum_network: NetworkId) -> Self {
+		Self {
+			iter: message.inner().iter().peekable(),
+			ethereum_network,
+			_marker: Default::default(),
+		}
+	}
+
+	pub fn convert(&mut self) -> Result<Message, XcmConverterError> {
+		let result = self.to_ethereum_message()?;
+		Ok(result)
+	}
+
+	fn next(&mut self) -> Result<&'a Instruction<Call>, XcmConverterError> {
+		self.iter.next().ok_or(XcmConverterError::UnexpectedEndOfXcm)
+	}
+
+	fn peek(&mut self) -> Result<&&'a Instruction<Call>, XcmConverterError> {
+		self.iter.peek().ok_or(XcmConverterError::UnexpectedEndOfXcm)
+	}
+
+	fn network_matches(&self, network: &Option<NetworkId>) -> bool {
+		if let Some(network) = network {
+			*network == self.ethereum_network
+		} else {
+			true
+		}
+	}
+
+	/// Extract the fee asset item from PayFees(V5)
+	fn extract_remote_fee(&mut self) -> Result<u128, XcmConverterError> {
+		use XcmConverterError::*;
+		let _ = match_expression!(self.next()?, WithdrawAsset(fee), fee)
+			.ok_or(WithdrawAssetExpected)?;
+		let fee_asset =
+			match_expression!(self.next()?, PayFees { asset: fee }, fee).ok_or(InvalidFeeAsset)?;
+		let (fee_asset_id, fee_amount) = match fee_asset {
+			Asset { id: asset_id, fun: Fungible(amount) } => Some((asset_id, *amount)),
+			_ => None,
+		}
+		.ok_or(AssetResolutionFailed)?;
+		// Check the fee asset is Ether
+		ensure!(fee_asset_id.0 == Here.into(), InvalidFeeAsset);
+		Ok(fee_amount)
+	}
+
+	/// Convert the xcm into the Message which will be executed
+	/// on Ethereum Gateway contract, we expect an input of the form:
+	/// # WithdrawAsset(WETH)
+	/// # PayFees(WETH)
+	/// # ReserveAssetDeposited(PNA) | WithdrawAsset(ENA)
+	/// # AliasOrigin(Origin)
+	/// # DepositAsset(PNA|ENA)
+	/// # Transact() ---Optional
+	/// # SetTopic
```

Comment:
@yrong lets make this documentation neater:

```suggestion
	/// Convert the xcm into an outbound message which can be dispatched to
	/// the Gateway contract on Ethereum
	///
	/// Assets being transferred can either be Polkadot-native assets (PNA)
	/// or Ethereum-native assets (ENA).
	/// 
	/// Expected Input Syntax: 
	/// ```ignore
	/// WithdrawAsset(ETH)
	/// PayFees(ETH)
	/// ReserveAssetDeposited(PNA) | WithdrawAsset(ENA)
	/// AliasOrigin(Origin)
	/// DepositAsset(Asset)
	/// Transact() [OPTIONAL]
	/// SetTopic(Topic)
	/// ```
```

Can you also please include in the documentation whether other instructions are ignored, or whether their presence cause conversion to fail.

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/d74a31ab0d4915b3ac57182605d6a67317241635